# OPTIMISATION 1: increase number of layers.

## Preprocessing

In [5]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf

#  Import and read the charity_data.csv.
import pandas as pd 
application_df = pd.read_csv("Resources/charity_data.csv")
application_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [2]:
# Drop the non-beneficial ID columns, 'EIN' and 'NAME'.
application_df = application_df.drop(columns=['EIN', 'NAME'])
application_df

APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
0                  T10       Independent          C1000    ProductDev   
1                   T3       Independent          C2000  Preservation   
2                   T5  CompanySponsored          C3000    ProductDev   
3                   T3  CompanySponsored          C2000  Preservation   
4                   T3       Independent          C1000     Heathcare   
...                ...               ...            ...           ...   
34294               T4       Independent          C1000    ProductDev   
34295               T4  CompanySponsored          C3000    ProductDev   
34296               T3  CompanySponsored          C2000  Preservation   
34297               T5       Independent          C3000    ProductDev   
34298               T3       Independent          C1000  Preservation   

       ORGANIZATION  STATUS     INCOME_AMT SPECIAL_CONSIDERATIONS   ASK_AMT  \
0       Association       1              0                      N      5000   
1      Co-operative       1         1-9999                      N    108590   
2       Association       1              0                      N      5000   
3             Trust       1    10000-24999                      N      6692   
4             Trust       1  100000-499999                      N    142590   
...             ...     ...            ...                    ...       ...   
34294   Association       1              0                      N      5000   
34295   Association       1              0                      N      5000   
34296   Association       1              0                      N      5000   
34297   Association       1              0                      N      5000   
34298  Co-operative       1          1M-5M                      N  36500179   

       IS_SUCCESSFUL  
0                  1  
1                  1  
2                  0  
3                  1  
4                  1  
...              ...  
34294              0  
34295              0  
34296              0  
34297              1  
34298              0  

[34299 rows x 10 columns]

In [3]:
# Determine the number of unique values in each column.
application_df.nunique()

APPLICATION_TYPE            17
AFFILIATION                  6
CLASSIFICATION              71
USE_CASE                     5
ORGANIZATION                 4
STATUS                       2
INCOME_AMT                   9
SPECIAL_CONSIDERATIONS       2
ASK_AMT                   8747
IS_SUCCESSFUL                2
dtype: int64

In [4]:
# Look at APPLICATION_TYPE value counts for binning
appSeries = application_df['APPLICATION_TYPE'].value_counts()

In [5]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`
app_cutoff = 500

application_types_to_replace = []

for item in appSeries.index:
    if appSeries[item]<app_cutoff:
        application_types_to_replace.append(item)

In [6]:
# Replace in dataframe
for app in application_types_to_replace:
    application_df['APPLICATION_TYPE'] = application_df['APPLICATION_TYPE'].replace(app,"Other")

# Check to make sure binning was successful
application_df['APPLICATION_TYPE'].value_counts()

T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
Other      276
Name: APPLICATION_TYPE, dtype: int64

In [7]:
# Look at CLASSIFICATION value counts for binning
classSeries = application_df['CLASSIFICATION'].value_counts()
classSeries

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: CLASSIFICATION, Length: 71, dtype: int64

In [8]:
# You may find it helpful to look at CLASSIFICATION value counts >1

#find list of values to drop
for item in classSeries.index:
    if classSeries[item] == 1:
        application_df = application_df[application_df.CLASSIFICATION != item]
        classSeries = classSeries.drop(item)
application_df['CLASSIFICATION'].value_counts()

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
C2800       95
C7100       75
C1300       58
C1280       50
C1230       36
C1400       34
C7200       32
C2300       32
C1240       30
C8000       20
C7120       18
C1500       16
C1800       15
C6000       15
C1250       14
C8200       11
C1238       10
C1278       10
C1237        9
C1235        9
C7210        7
C4100        6
C1720        6
C2400        6
C1600        5
C1257        5
C1260        3
C2710        3
C0           3
C1267        2
C1256        2
C1234        2
C1246        2
C3200        2
Name: CLASSIFICATION, dtype: int64

In [9]:
# Choose a cutoff value and create a list of classifications to be replaced
# use the variable name `classifications_to_replace`
class_cutoff = 100

classifications_to_replace= []

for item in classSeries.index:
    if classSeries[item]<class_cutoff:
        classifications_to_replace.append(item)
        
classifications_to_replace 

['C2800',
 'C7100',
 'C1300',
 'C1280',
 'C1230',
 'C1400',
 'C7200',
 'C2300',
 'C1240',
 'C8000',
 'C7120',
 'C1500',
 'C1800',
 'C6000',
 'C1250',
 'C8200',
 'C1238',
 'C1278',
 'C1235',
 'C1237',
 'C7210',
 'C2400',
 'C1720',
 'C4100',
 'C1257',
 'C1600',
 'C1260',
 'C2710',
 'C0',
 'C3200',
 'C1234',
 'C1246',
 'C1267',
 'C1256']

In [10]:
# Replace in dataframe
for cls in classifications_to_replace:
    application_df['CLASSIFICATION'] = application_df['CLASSIFICATION'].replace(cls,"Other")
    
# Check to make sure binning was successful
application_df['CLASSIFICATION'].value_counts()

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
Other      643
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
Name: CLASSIFICATION, dtype: int64

In [11]:
# Convert categorical data to numeric with `pd.get_dummies`

# What are the data types in the df?
type_list = application_df.dtypes

In [12]:
# create a list of columns to convert
for item in type_list.index:
    if type_list[item] == 'int64':
        item
        type_list = type_list.drop(item)

In [13]:
application_df = pd.get_dummies(application_df, columns=type_list.index)

In [14]:
# Split our preprocessed data into our features and target arrays
y = application_df.IS_SUCCESSFUL.values
X = application_df.drop(columns="IS_SUCCESSFUL").values

# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42, stratify=y)

In [15]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

In [16]:
input_dimensions=X.shape[1]
input_dimensions

49

## Compile, Train and Evaluate the Model

In [17]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.

nn_model = tf.keras.models.Sequential()

nn_model.add(tf.keras.layers.Dense(units=8, activation="relu", input_dim=input_dimensions))

# First hidden layer
nn_model.add(tf.keras.layers.Dense(units=8, activation="relu"))

# Second hidden layer
nn_model.add(tf.keras.layers.Dense(units=8, activation="relu"))

# Third hidden layer
nn_model.add(tf.keras.layers.Dense(units=8, activation="relu"))

# Output layer
nn_model.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn_model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 8)                 400       
                                                                 
 dense_1 (Dense)             (None, 8)                 72        
                                                                 
 dense_2 (Dense)             (None, 8)                 72        
                                                                 
 dense_3 (Dense)             (None, 8)                 72        
                                                                 
 dense_4 (Dense)             (None, 1)                 9         
                                                                 
Total params: 625
Trainable params: 625
Non-trainable params: 0
_________________________________________________________________


In [18]:
# Compile the model
nn_model.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [19]:
# Train the model
fit_model = nn_model.fit(X_train_scaled, y_train, epochs=100)

Epoch 1/100
804/804 [==============================] - 3s 2ms/step - loss: 0.6009 - accuracy: 0.6921
Epoch 2/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5589 - accuracy: 0.7277
Epoch 3/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5533 - accuracy: 0.7289
Epoch 4/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5511 - accuracy: 0.7311
Epoch 5/100
804/804 [==============================] - 3s 3ms/step - loss: 0.5491 - accuracy: 0.7305
Epoch 6/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5473 - accuracy: 0.7314
Epoch 7/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5467 - accuracy: 0.7337
Epoch 8/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5454 - accuracy: 0.7341
Epoch 9/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5450 - accuracy: 0.7342
Epoch 10/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5447 - accura

In [20]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn_model.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 0s - loss: 0.5602 - accuracy: 0.7239 - 384ms/epoch - 1ms/step
Loss: 0.560231626033783, Accuracy: 0.7238884568214417


# The model is still only achieving 72.4% accuracy, short of the 75% target.

# Optimisation 2: increase number of neurons in layers.

## Compile, Train and Evaluate the Model

In [21]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.

nn_model = tf.keras.models.Sequential()

nn_model.add(tf.keras.layers.Dense(units=16, activation="relu", input_dim=input_dimensions))

# First hidden layer
nn_model.add(tf.keras.layers.Dense(units=16, activation="relu"))

# Second hidden layer
nn_model.add(tf.keras.layers.Dense(units=16, activation="relu"))

# Third hidden layer
nn_model.add(tf.keras.layers.Dense(units=16, activation="relu"))

# Output layer
nn_model.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn_model.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_5 (Dense)             (None, 16)                800       
                                                                 
 dense_6 (Dense)             (None, 16)                272       
                                                                 
 dense_7 (Dense)             (None, 16)                272       
                                                                 
 dense_8 (Dense)             (None, 16)                272       
                                                                 
 dense_9 (Dense)             (None, 1)                 17        
                                                                 
Total params: 1,633
Trainable params: 1,633
Non-trainable params: 0
_________________________________________________________________


In [22]:
# Compile the model
nn_model.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [23]:
# Train the model
fit_model = nn_model.fit(X_train_scaled, y_train, epochs=100)

Epoch 1/100
804/804 [==============================] - 2s 1ms/step - loss: 0.5813 - accuracy: 0.7117
Epoch 2/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5532 - accuracy: 0.7305
Epoch 3/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5491 - accuracy: 0.7332
Epoch 4/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5474 - accuracy: 0.7330
Epoch 5/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5455 - accuracy: 0.7355
Epoch 6/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5442 - accuracy: 0.7348
Epoch 7/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5431 - accuracy: 0.7358
Epoch 8/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5434 - accuracy: 0.7365
Epoch 9/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5424 - accuracy: 0.7376
Epoch 10/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5416 - accura

In [24]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn_model.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 0s - loss: 0.5667 - accuracy: 0.7233 - 325ms/epoch - 1ms/step
Loss: 0.5666735768318176, Accuracy: 0.7233049273490906


# The model is still only achieving 72.3% accuracy, short of the 75% target.

# Optimisation 3: adjusting neuron functions.

In [25]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.

nn_model = tf.keras.models.Sequential()

nn_model.add(tf.keras.layers.Dense(units=16, activation="tanh", input_dim=input_dimensions))

# First hidden layer
nn_model.add(tf.keras.layers.Dense(units=16, activation="tanh"))

# Second hidden layer
nn_model.add(tf.keras.layers.Dense(units=16, activation="tanh"))

# Third hidden layer
nn_model.add(tf.keras.layers.Dense(units=16, activation="tanh"))

# Output layer
nn_model.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn_model.summary()

Model: "sequential_2"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_10 (Dense)            (None, 16)                800       
                                                                 
 dense_11 (Dense)            (None, 16)                272       
                                                                 
 dense_12 (Dense)            (None, 16)                272       
                                                                 
 dense_13 (Dense)            (None, 16)                272       
                                                                 
 dense_14 (Dense)            (None, 1)                 17        
                                                                 
Total params: 1,633
Trainable params: 1,633
Non-trainable params: 0
_________________________________________________________________


In [26]:
# Compile the model
nn_model.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [27]:
# Train the model
fit_model = nn_model.fit(X_train_scaled, y_train, epochs=100)

Epoch 1/100
804/804 [==============================] - 2s 1ms/step - loss: 0.5801 - accuracy: 0.7069
Epoch 2/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5550 - accuracy: 0.7284
Epoch 3/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5521 - accuracy: 0.7322
Epoch 4/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5497 - accuracy: 0.7314
Epoch 5/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5485 - accuracy: 0.7330
Epoch 6/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5470 - accuracy: 0.7336
Epoch 7/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5466 - accuracy: 0.7348
Epoch 8/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5457 - accuracy: 0.7344
Epoch 9/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5444 - accuracy: 0.7354
Epoch 10/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5439 - accura

In [28]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn_model.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 0s - loss: 0.5625 - accuracy: 0.7217 - 334ms/epoch - 1ms/step
Loss: 0.5624932646751404, Accuracy: 0.7216711640357971


# The model is still only achieving 72.2% accuracy, short of the 75% target.

# Optimisation 4: re-processing data and Auto-tuning hyperparameters.

In [6]:

application_df = pd.read_csv("Resources/charity_data.csv")
application_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [7]:
# Drop the non-beneficial ID columns, 'EIN' and 'NAME'.
application_df = application_df.drop(columns=['EIN', 'NAME'])
application_df

APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
0                  T10       Independent          C1000    ProductDev   
1                   T3       Independent          C2000  Preservation   
2                   T5  CompanySponsored          C3000    ProductDev   
3                   T3  CompanySponsored          C2000  Preservation   
4                   T3       Independent          C1000     Heathcare   
...                ...               ...            ...           ...   
34294               T4       Independent          C1000    ProductDev   
34295               T4  CompanySponsored          C3000    ProductDev   
34296               T3  CompanySponsored          C2000  Preservation   
34297               T5       Independent          C3000    ProductDev   
34298               T3       Independent          C1000  Preservation   

       ORGANIZATION  STATUS     INCOME_AMT SPECIAL_CONSIDERATIONS   ASK_AMT  \
0       Association       1              0                      N      5000   
1      Co-operative       1         1-9999                      N    108590   
2       Association       1              0                      N      5000   
3             Trust       1    10000-24999                      N      6692   
4             Trust       1  100000-499999                      N    142590   
...             ...     ...            ...                    ...       ...   
34294   Association       1              0                      N      5000   
34295   Association       1              0                      N      5000   
34296   Association       1              0                      N      5000   
34297   Association       1              0                      N      5000   
34298  Co-operative       1          1M-5M                      N  36500179   

       IS_SUCCESSFUL  
0                  1  
1                  1  
2                  0  
3                  1  
4                  1  
...              ...  
34294              0  
34295              0  
34296              0  
34297              1  
34298              0  

[34299 rows x 10 columns]

In [8]:
# define a function to remove noisy data, due to repetitive coding

def noise_cleanup(d8aframe, col, cutoff):
    column_series = d8aframe[col].value_counts()
    
    for item in column_series.index:
        if column_series[item]<cutoff:
            print(f"{item} is going down!")
            column_series = column_series.drop(item)
            d8aframe = d8aframe[d8aframe[col] != item]
    print()
    print("data remaining:")
    print(d8aframe[col].value_counts())
    
    return d8aframe

In [9]:
# Determine the number of unique values in each column.
application_df.nunique()

APPLICATION_TYPE            17
AFFILIATION                  6
CLASSIFICATION              71
USE_CASE                     5
ORGANIZATION                 4
STATUS                       2
INCOME_AMT                   9
SPECIAL_CONSIDERATIONS       2
ASK_AMT                   8747
IS_SUCCESSFUL                2
dtype: int64

In [10]:
# check each column to see how noisy it is, to make an informed decision on how much to simplify the data.

application_df.APPLICATION_TYPE.value_counts()

T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: APPLICATION_TYPE, dtype: int64

In [11]:
# drop anything less than 100
application_df = noise_cleanup(application_df, 'APPLICATION_TYPE', 100)

T13 is going down!
T12 is going down!
T2 is going down!
T25 is going down!
T14 is going down!
T29 is going down!
T15 is going down!
T17 is going down!

data remaining:
T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
Name: APPLICATION_TYPE, dtype: int64


In [12]:
application_df.AFFILIATION.value_counts()

Independent         18364
CompanySponsored    15703
Family/Parent          64
National               32
Regional               12
Other                   4
Name: AFFILIATION, dtype: int64

In [13]:
# drop anything less than 100
application_df = noise_cleanup(application_df, 'AFFILIATION', 100)

Family/Parent is going down!
National is going down!
Regional is going down!
Other is going down!

data remaining:
Independent         18364
CompanySponsored    15703
Name: AFFILIATION, dtype: int64


In [14]:
application_df.CLASSIFICATION.value_counts()

C1000    17160
C2000     6034
C1200     4824
C3000     1917
C2100     1881
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: CLASSIFICATION, Length: 71, dtype: int64

In [15]:
application_df = noise_cleanup(application_df, 'CLASSIFICATION', 100)

C2800 is going down!
C7100 is going down!
C1300 is going down!
C1280 is going down!
C1230 is going down!
C1400 is going down!
C7200 is going down!
C2300 is going down!
C1240 is going down!
C8000 is going down!
C7120 is going down!
C1500 is going down!
C1800 is going down!
C6000 is going down!
C1250 is going down!
C8200 is going down!
C1238 is going down!
C1278 is going down!
C1235 is going down!
C1237 is going down!
C7210 is going down!
C2400 is going down!
C1720 is going down!
C4100 is going down!
C1257 is going down!
C1600 is going down!
C1260 is going down!
C2710 is going down!
C0 is going down!
C3200 is going down!
C1234 is going down!
C1246 is going down!
C1267 is going down!
C1256 is going down!
C2190 is going down!
C4200 is going down!
C2600 is going down!
C5200 is going down!
C1370 is going down!
C1248 is going down!
C6100 is going down!
C1820 is going down!
C1900 is going down!
C1236 is going down!
C3700 is going down!
C2570 is going down!
C1580 is going down!
C1245 is going d

In [16]:
application_df.USE_CASE.value_counts()

Preservation     27330
ProductDev        5547
CommunityServ      379
Heathcare          142
Other                3
Name: USE_CASE, dtype: int64

In [17]:
application_df = noise_cleanup(application_df, 'USE_CASE', 100)

Other is going down!

data remaining:
Preservation     27330
ProductDev        5547
CommunityServ      379
Heathcare          142
Name: USE_CASE, dtype: int64


In [18]:
application_df.ORGANIZATION.value_counts()

Trust           22850
Association     10065
Co-operative      445
Corporation        38
Name: ORGANIZATION, dtype: int64

In [19]:
application_df = noise_cleanup(application_df, 'ORGANIZATION', 100)

Corporation is going down!

data remaining:
Trust           22850
Association     10065
Co-operative      445
Name: ORGANIZATION, dtype: int64


In [20]:
application_df.STATUS.value_counts()

1    33355
0        5
Name: STATUS, dtype: int64

In [21]:
application_df = noise_cleanup(application_df, 'STATUS', 100)

0 is going down!

data remaining:
1    33355
Name: STATUS, dtype: int64


### I will drop columns like STATUS from the dataframe as they are no longer variables and have no function.

In [22]:
application_df = application_df.drop(columns=["STATUS"])

In [23]:
application_df.SPECIAL_CONSIDERATIONS.value_counts()

N    33329
Y       26
Name: SPECIAL_CONSIDERATIONS, dtype: int64

In [24]:
# drop this column as there are not enough Y values to be significant.
application_df = application_df.drop(columns=["SPECIAL_CONSIDERATIONS"])

## Group ASK_AMT and INCOME_AMT into fewer bins to reduce noise

In [25]:
application_df.ASK_AMT.value_counts()

5000        24751
63981           3
15583           3
6725            3
10478           3
            ...  
44686           1
55308           1
432239          1
37365916        1
36500179        1
Name: ASK_AMT, Length: 8461, dtype: int64

In [26]:
# have a look at the range of the data to help find useful bin ranges...
application_df.ASK_AMT.describe()

count    3.335500e+04
mean     2.652789e+06
std      8.742741e+07
min      5.000000e+03
25%      5.000000e+03
50%      5.000000e+03
75%      7.143500e+03
max      8.597806e+09
Name: ASK_AMT, dtype: float64

In [27]:
ask_cutoff = 100000

ask_Series = application_df['ASK_AMT'].value_counts()
ask_Series

5000        24751
63981           3
15583           3
6725            3
10478           3
            ...  
44686           1
55308           1
432239          1
37365916        1
36500179        1
Name: ASK_AMT, Length: 8461, dtype: int64

In [28]:
ask_Series = ask_Series.drop(5000)
ask_Series

63981       3
15583       3
6725        3
10478       3
5706        2
           ..
44686       1
55308       1
432239      1
37365916    1
36500179    1
Name: ASK_AMT, Length: 8460, dtype: int64

In [29]:
ask_low = []
ask_high = []

for item in ask_Series.index:
    if item < ask_cutoff:
        ask_low.append(item)
    else:
        ask_high.append(item)
        

In [30]:
ask_low

[63981,
 15583,
 6725,
 10478,
 5706,
 14000,
 8632,
 6398,
 67367,
 12584,
 9537,
 7686,
 43814,
 5130,
 86039,
 39168,
 18063,
 5297,
 61397,
 27944,
 16200,
 27159,
 13898,
 17614,
 22579,
 7187,
 42072,
 5731,
 13420,
 6742,
 39122,
 80599,
 5179,
 17010,
 31549,
 59647,
 24459,
 83827,
 14765,
 13499,
 18985,
 34995,
 90834,
 5470,
 13831,
 50890,
 54024,
 36683,
 18826,
 73627,
 9323,
 32507,
 46327,
 21264,
 21389,
 10070,
 31453,
 34875,
 20834,
 30769,
 73987,
 40743,
 32807,
 20537,
 5255,
 57588,
 19704,
 40429,
 21346,
 16435,
 7003,
 6279,
 47129,
 83211,
 16482,
 6010,
 20071,
 9413,
 29541,
 45793,
 23202,
 26432,
 11839,
 5170,
 27441,
 42178,
 6035,
 35734,
 5175,
 13681,
 20150,
 27073,
 32141,
 16071,
 63520,
 6920,
 80168,
 60000,
 9255,
 34792,
 64093,
 8746,
 6960,
 5006,
 6405,
 94253,
 56135,
 24746,
 41894,
 5894,
 7049,
 17735,
 37390,
 22472,
 15372,
 6000,
 37489,
 57689,
 9885,
 12238,
 25952,
 12753,
 57957,
 34538,
 20684,
 15039,
 8181,
 21068,
 66909,
 

In [31]:
ask_high

[184668,
 136341,
 240780,
 195661,
 238933,
 146824,
 246851,
 201112,
 150599,
 206725,
 100000,
 159223,
 210432,
 10069081,
 133556,
 7903843,
 11002681,
 1477563,
 216783,
 524569,
 141851,
 203256,
 119499,
 912811,
 393444,
 164860,
 164874,
 247827,
 161531,
 438018,
 130381,
 245637,
 802010,
 210098,
 24766071,
 94492274,
 180152,
 3189005,
 580496,
 1052902,
 45165663,
 220526,
 950074,
 15259643,
 297301,
 107588,
 177811,
 1199736,
 440617,
 1586920,
 4510717,
 173209,
 2811429,
 1519567,
 2731896,
 127793,
 1124121,
 237483,
 120304,
 2192625,
 9721128,
 2168130,
 107944,
 512014,
 111480,
 6201027,
 985895,
 287765,
 131391,
 1189341,
 1112173,
 1018486,
 7429956,
 586682,
 6074233,
 690720,
 105721,
 689619,
 10395185,
 106141,
 615128,
 693028,
 283821,
 2766059,
 455438,
 1407719,
 590133,
 109088,
 282870,
 922986,
 135185,
 112883,
 2731369,
 291470,
 568256,
 1150846,
 20606812,
 306415,
 143601,
 147193,
 338547,
 132892,
 1204846,
 512542,
 7992807,
 1687740,
 12

In [32]:
# Replace in dataframe
for ask in ask_low:
    application_df['ASK_AMT'] = application_df['ASK_AMT'].replace(ask,"5,000 - 100,000")
    
for ask in ask_high:
    application_df['ASK_AMT'] = application_df['ASK_AMT'].replace(ask,"greater than 100,000")

# Check to make sure binning was successful
application_df['ASK_AMT'].value_counts()

5000                    24751
greater than 100,000     4351
5,000 - 100,000          4253
Name: ASK_AMT, dtype: int64

In [33]:
application_df.INCOME_AMT.value_counts()

0                23770
25000-99999       3652
100000-499999     3260
1M-5M              918
1-9999             696
10000-24999        523
10M-50M            235
5M-10M             175
50M+               126
Name: INCOME_AMT, dtype: int64

In [34]:
# will have to hand code this due to text-based categories

income_low = ["1-9999", "10000-24999", "25000-99999", "25000-99999", "100000-499999"]
income_high = ["1M-5M", "5M-10M", "10M-50M", "50M+"]

In [35]:
# Replace in dataframe
for income in income_low:
    application_df['INCOME_AMT'] = application_df['INCOME_AMT'].replace(income,"1-1,000,000")
    
for income in income_high:
    application_df['INCOME_AMT'] = application_df['INCOME_AMT'].replace(income,"greater than 1,000,000")

# Check to make sure binning was successful
application_df['INCOME_AMT'].value_counts()

0                         23770
1-1,000,000                8131
greater than 1,000,000     1454
Name: INCOME_AMT, dtype: int64

In [36]:
# Convert categorical data to numeric with `pd.get_dummies`

# What are the data types in the df?
type_list = application_df.dtypes
type_list

APPLICATION_TYPE    object
AFFILIATION         object
CLASSIFICATION      object
USE_CASE            object
ORGANIZATION        object
INCOME_AMT          object
ASK_AMT             object
IS_SUCCESSFUL        int64
dtype: object

In [37]:
# create a list of columns to convert
for item in type_list.index:
    if type_list[item] == 'int64':
        item
        type_list = type_list.drop(item)

In [38]:
application_df

APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
0                  T10       Independent          C1000    ProductDev   
1                   T3       Independent          C2000  Preservation   
2                   T5  CompanySponsored          C3000    ProductDev   
3                   T3  CompanySponsored          C2000  Preservation   
4                   T3       Independent          C1000     Heathcare   
...                ...               ...            ...           ...   
34294               T4       Independent          C1000    ProductDev   
34295               T4  CompanySponsored          C3000    ProductDev   
34296               T3  CompanySponsored          C2000  Preservation   
34297               T5       Independent          C3000    ProductDev   
34298               T3       Independent          C1000  Preservation   

       ORGANIZATION              INCOME_AMT               ASK_AMT  \
0       Association                       0                  5000   
1      Co-operative             1-1,000,000  greater than 100,000   
2       Association                       0                  5000   
3             Trust             1-1,000,000       5,000 - 100,000   
4             Trust             1-1,000,000  greater than 100,000   
...             ...                     ...                   ...   
34294   Association                       0                  5000   
34295   Association                       0                  5000   
34296   Association                       0                  5000   
34297   Association                       0                  5000   
34298  Co-operative  greater than 1,000,000  greater than 100,000   

       IS_SUCCESSFUL  
0                  1  
1                  1  
2                  0  
3                  1  
4                  1  
...              ...  
34294              0  
34295              0  
34296              0  
34297              1  
34298              0  

[33355 rows x 8 columns]

In [39]:
application_df = pd.get_dummies(application_df, columns=type_list.index)
application_df

IS_SUCCESSFUL  APPLICATION_TYPE_T10  APPLICATION_TYPE_T19  \
0                  1                     1                     0   
1                  1                     0                     0   
2                  0                     0                     0   
3                  1                     0                     0   
4                  1                     0                     0   
...              ...                   ...                   ...   
34294              0                     0                     0   
34295              0                     0                     0   
34296              0                     0                     0   
34297              1                     0                     0   
34298              0                     0                     0   

       APPLICATION_TYPE_T3  APPLICATION_TYPE_T4  APPLICATION_TYPE_T5  \
0                        0                    0                    0   
1                        1                    0                    0   
2                        0                    0                    1   
3                        1                    0                    0   
4                        1                    0                    0   
...                    ...                  ...                  ...   
34294                    0                    1                    0   
34295                    0                    1                    0   
34296                    1                    0                    0   
34297                    0                    0                    1   
34298                    1                    0                    0   

       APPLICATION_TYPE_T6  APPLICATION_TYPE_T7  APPLICATION_TYPE_T8  \
0                        0                    0                    0   
1                        0                    0                    0   
2                        0                    0                    0   
3                        0                    0                    0   
4                        0                    0                    0   
...                    ...                  ...                  ...   
34294                    0                    0                    0   
34295                    0                    0                    0   
34296                    0                    0                    0   
34297                    0                    0                    0   
34298                    0                    0                    0   

       APPLICATION_TYPE_T9  ...  USE_CASE_ProductDev  \
0                        0  ...                    1   
1                        0  ...                    0   
2                        0  ...                    1   
3                        0  ...                    0   
4                        0  ...                    0   
...                    ...  ...                  ...   
34294                    0  ...                    1   
34295                    0  ...                    1   
34296                    0  ...                    0   
34297                    0  ...                    1   
34298                    0  ...                    0   

       ORGANIZATION_Association  ORGANIZATION_Co-operative  \
0                             1                          0   
1                             0                          1   
2                             1                          0   
3                             0                          0   
4                             0                          0   
...                         ...                        ...   
34294                         1                          0   
34295                         1                          0   
34296                         1                          0   
34297                         1                          0   
34298                         0                          1   

       ORGANIZATION_Trust  INCOME_AMT_0  I

In [40]:
# Split our preprocessed data into our features and target arrays
y = application_df.IS_SUCCESSFUL.values
X = application_df.drop(columns="IS_SUCCESSFUL").values

# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42, stratify=y)

In [41]:
input_dimensions=X.shape[1]
input_dimensions

35

In [42]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

In [43]:
# Create a method that creates a new Sequential model with hyperparameter options
def create_model(hp):
    nn_model = tf.keras.models.Sequential()

    # Allow kerastuner to decide which activation function to use in hidden layers
    activation = hp.Choice('activation',['relu','tanh'])
    
    # Allow kerastuner to decide number of neurons in first layer
    nn_model.add(tf.keras.layers.Dense(units=hp.Int('first_units',
        min_value=1,
        max_value=80,
        step=5), activation=activation, input_dim=input_dimensions))

    # Allow kerastuner to decide number of hidden layers and neurons in hidden layers
    for i in range(hp.Int('num_layers', 1, 5)):
        nn_model.add(tf.keras.layers.Dense(units=hp.Int('units_' + str(i),
            min_value=1,
            max_value=100,
            step=5),
            activation=activation))
    
    nn_model.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

    # Compile the model
    nn_model.compile(loss="binary_crossentropy", optimizer='adam', metrics=["accuracy"])
    
    return nn_model

In [45]:
# Import the kerastuner library
import keras_tuner as kt

tuner = kt.Hyperband(
    create_model,
    objective="val_accuracy",
    overwrite=True,
    max_epochs=20,
    hyperband_iterations=2,)

In [46]:
# Run the kerastuner search for best hyperparameters
tuner.search(X_train_scaled,y_train,epochs=20,validation_data=(X_test_scaled,y_test))

Trial 60 Complete [00h 00m 29s]
val_accuracy: 0.7251468896865845

Best val_accuracy So Far: 0.7261062264442444
Total elapsed time: 00h 11m 12s
INFO:tensorflow:Oracle triggered exit


In [ ]:
# Get best model hyperparameters
best_hyper = tuner.get_best_hyperparameters(1)[0]
best_hyper.values

In [ ]:
# Evaluate best model against full test data
best_model = tuner.get_best_models(1)[0]
model_loss, model_accuracy = best_model.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

In [ ]:
# Export our model to HDF5 file
best_model.save('AlphabetSoupCharity_Optimization.h5')


## Despite an extensive search, I have not managed higher than 72.6% accuracy.